dot/core/service_test.go

```diff
@@ -577,12 +657,708 @@ func Test_Service_handleBlocksAsync(t *testing.T) {
+			blockState: mockBlockState,
+		}
+		execTest(t, service, testPrevHash, testCurrentHash, errDummyErr)
+	})
+
+	t.Run("empty subchain", func(t *testing.T) {
+		t.Parallel()
+		ctrl := gomock.NewController(t)
+		mockBlockState := NewMockBlockState(ctrl)
+		mockBlockState.EXPECT().HighestCommonAncestor(testPrevHash, testCurrentHash).
+			Return(testAncestorHash, nil)
+		mockBlockState.EXPECT().SubChain(testAncestorHash, testPrevHash).Return([]common.Hash{}, nil)
+
+		service := &Service{
+			blockState: mockBlockState,
+		}
+		execTest(t, service, testPrevHash, testCurrentHash, nil)
+	})
+
+	t.Run("get runtime err", func(t *testing.T) {
+		t.Parallel()
+		ctrl := gomock.NewController(t)
+		mockBlockState := NewMockBlockState(ctrl)
+		mockBlockState.EXPECT().HighestCommonAncestor(testPrevHash, testCurrentHash).
+			Return(testAncestorHash, nil)
+		mockBlockState.EXPECT().SubChain(testAncestorHash, testPrevHash).Return(testSubChain, nil)
+		mockBlockState.EXPECT().GetRuntime(nil).Return(nil, errDummyErr)
+
+		service := &Service{
+			blockState: mockBlockState,
+		}
+		execTest(t, service, testPrevHash, testCurrentHash, errDummyErr)
+	})
+
+	t.Run("invalid transaction", func(t *testing.T) {
+		t.Parallel()
+		ctrl := gomock.NewController(t)
+		runtimeMockErr := new(mocksruntime.Instance)
+		mockBlockState := NewMockBlockState(ctrl)
+		mockBlockState.EXPECT().HighestCommonAncestor(testPrevHash, testCurrentHash).
+			Return(testAncestorHash, nil)
+		mockBlockState.EXPECT().SubChain(testAncestorHash, testPrevHash).Return(testSubChain, nil)
+		mockBlockState.EXPECT().GetRuntime(nil).Return(runtimeMockErr, nil)
+		mockBlockState.EXPECT().GetBlockBody(testCurrentHash).Return(nil, errDummyErr)
+		mockBlockState.EXPECT().GetBlockBody(testAncestorHash).Return(body, nil)
+		runtimeMockErr.On("ValidateTransaction", externExt).Return(nil, errTestDummyError)
+
+		service := &Service{
+			blockState: mockBlockState,
+		}
+		execTest(t, service, testPrevHash, testCurrentHash, nil)
+	})
+
+	t.Run("happy path", func(t *testing.T) {
+		t.Parallel()
+		ctrl := gomock.NewController(t)
+		runtimeMockOk := new(mocksruntime.Instance)
+		mockBlockState := NewMockBlockState(ctrl)
+		mockBlockState.EXPECT().HighestCommonAncestor(testPrevHash, testCurrentHash).
+			Return(testAncestorHash, nil)
+		mockBlockState.EXPECT().SubChain(testAncestorHash, testPrevHash).Return(testSubChain, nil)
+		mockBlockState.EXPECT().GetRuntime(nil).Return(runtimeMockOk, nil)
+		mockBlockState.EXPECT().GetBlockBody(testCurrentHash).Return(nil, errDummyErr)
+		mockBlockState.EXPECT().GetBlockBody(testAncestorHash).Return(body, nil)
+		runtimeMockOk.On("ValidateTransaction", externExt).
+			Return(testValidity, nil)
+		mockTxnStateOk := NewMockTransactionState(ctrl)
+		mockTxnStateOk.EXPECT().AddToPool(vtx).Return(common.Hash{})
+
+		service := &Service{
+			blockState:       mockBlockState,
+			transactionState: mockTxnStateOk,
+		}
+		execTest(t, service, testPrevHash, testCurrentHash, nil)
+	})
```

Comment:
nit this is fine, although it might a bit cleaner with mock configuration functions as test case fields ([example](https://gist.github.com/qdm12/99358e2529a5887ae13a936a7f1c8ae5#file-potato_test-go-L234))